flang/include/flang/Lower/CharacterExpr.h

```diff
@@ -159,15 +159,19 @@ class CharacterExprHelper {
   /// above instead of fixing it when needed.
   fir::ExtendedValue cleanUpCharacterExtendedValue(const fir::ExtendedValue &);
 
+  /// Create fir.char<k> singleton from \p code integer value.
+  mlir::Value createSingletonFromCode(mlir::Value code, int kind);
+  /// Returns integer value held in a character singleton.
+  mlir::Value extractCodeFromSingleton(mlir::Value singleton);
+
 private:
   /// FIXME: the implementation also needs a clean-up now that
   /// CharBoxValue are better propagated.
   fir::CharBoxValue materializeValue(mlir::Value str);
-  mlir::Type getReferenceType(const fir::CharBoxValue &c) const;
-  mlir::Type getReferenceType(mlir::Value str) const;
-  mlir::Type getSeqTy(const fir::CharBoxValue &c) const;
-  mlir::Type getSeqTy(mlir::Value str) const;
+  mlir::Type getUnknownLenType(const fir::CharBoxValue &c) const;
```

Comment:
Does this imply the LEN from the box is overridden with `?` ?

Reply:
The box was not changed, this returns a `fir.char<kind, ?>` type with the same kind as the provided box.
I removed it since it is not that useful of a helper here after adding `fir::CharacterType::getUnknowLen(kind)`.